Source/Core/VideoCommon/PixelShaderGen.cpp

```diff
@@ -732,8 +752,135 @@ uint WrapCoord(int coord, uint wrap, int size) {{
   }
 }
 
+void WriteCustomShaderStructImpl(ShaderCode* out, u32 num_stages, bool per_pixel_lighting,
+                                 const pixel_shader_uid_data* uid_data)
+{
+  out->Write("\tCustomShaderData custom_data;\n");
+
+  if (per_pixel_lighting)
+  {
+    out->Write("\tcustom_data.position = WorldPos;\n");
+    out->Write("\tcustom_data.normal = Normal;\n");
+  }
+  else
+  {
+    out->Write("\tcustom_data.position = float3(0, 0, 0);\n");
+    out->Write("\tcustom_data.normal = float3(0, 0, 0);\n");
+  }
+
+  if (uid_data->genMode_numtexgens == 0) [[unlikely]]
+  {
+    out->Write("\tcustom_data.texcoord[0] = float3(0, 0, 0);\n");
+  }
+  else
+  {
+    for (u32 i = 0; i < uid_data->genMode_numtexgens; ++i)
+    {
+      out->Write("\tif (tex{0}.z == 0.0)\n", i);
+      out->Write("\t{{\n");
+      out->Write("\t\tcustom_data.texcoord[{0}] = tex{0};\n", i);
+      out->Write("\t}}\n");
+      out->Write("\telse {{\n");
+      out->Write("\t\tcustom_data.texcoord[{0}] = float3(tex{0}.xy / tex{0}.z, 0);\n", i);
+      out->Write("\t}}\n");
+    }
+  }
+
+  for (u32 i = 0; i < uid_data->genMode_numindstages; ++i)
+  {
+    if ((uid_data->nIndirectStagesUsed & (1U << i)) != 0)
+    {
+      u32 texcoord = uid_data->GetTevindirefCoord(i);
+
+      // Quirk: when the tex coord is not less than the number of tex gens (i.e. the tex coord
+      // does not exist), then tex coord 0 is used (though sometimes glitchy effects happen on
+      // console). This affects the Mario portrait in Luigi's Mansion, where the developers forgot
+      // to set the number of tex gens to 2 (bug 11462).
+      if (texcoord >= uid_data->genMode_numtexgens)
+        texcoord = 0;
+
+      out->Write("\t{{\n");
+      out->Write("\t\tint2 fixpoint_uv = int2(custom_data.texcoord[{}].xy", texcoord);
+      out->Write(" * float2(" I_TEXDIMS "[{}].zw * 128));\n", texcoord);
+      out->Write("\t\tint2 tempcoord = fixpoint_uv >> " I_INDTEXSCALE "[{}].{};\n", i / 2,
+                 (i & 1) ? "zw" : "xy");
+      out->Write("\t\tcustom_data.texcoord[{0}] = float3(tempcoord / float2(" I_TEXDIMS
+                 "[{0}].zw * 128), 0);\n",
+                 texcoord);
+      out->Write("\t}}\n");
+    }
+  }
+  out->Write("\tcustom_data.texcoord_count = {};\n", uid_data->genMode_numtexgens);
+
+  for (u32 i = 0; i < 8; i++)
+  {
+    // Shader compilation complains if every index isn't initialized
+    out->Write("\tcustom_data.texmap_to_texcoord_index[{0}] = {0};\n", i);
+  }
+
+  // Try and do a best guess on what the texcoord index is
```

Comment:
> Indirect textures don't get mapped to anything here, so they would lack an entry in texmap_to_texcoord_index. This can be fixed probably.

Yeah, if you have any ideas here, please let me know :)

> If multiple textures both use the same texture coordinate, then only the later texture coordinate is used. I don't think there's any way to fix this.

Like mentioned, if they are the same, I don't think this matters?  Maybe I am misunderstanding you.

--

Regardless, these are great points to make.  I will add them.  Thanks!

Reply:
Ah, right, I rethought this, and multiple textures using the same texture coordinate won't cause an issue (you'd just have e.g. `texmap_to_texcoord_index[0] = 1` and `texmap_to_texcoord_index[1] = 1` which is fine).

Instead, there is an issue if the same texture is used multiple times within a single draw. For instance, this would affect bumpmapping in Rogue Squadron ([RogueSquadron2BumpMapping](https://fifo.ci/media/dff/RogueSquadron2BumpMapping.dff)), they sample texture 4 (which is a wavy greyscale texture that acts as a height for the sand, see object 12/offset 00025aca) twice, once with texture coordinate 4 and then again with texture coordinate 5, where texture coordinate 5 is generated by adding an offset to texture coordinate 4). With the current implementation, you'll get `texmap_to_texcoord_index[4] = 5`, but I don't think there's really a better way of representing this.

Reply:
For indirect textures, you'd want this:

```C++
for (u32 i = 0; i < uid_data->genMode_numindstages; ++i)
{
  if ((uid_data->nIndirectStagesUsed & (1U << i)) != 0)
  {
    u32 texcoord = uid_data->GetTevindirefCoord(i);
    const u32 texmap = uid_data->GetTevindirefMap(i);

    // Quirk: when the tex coord is not less than the number of tex gens (i.e. the tex coord does
    // not exist), then tex coord 0 is used (though sometimes glitchy effects happen on console).
    // This affects the Mario portrait in Luigi's Mansion, where the developers forgot to set
    // the number of tex gens to 2 (bug 11462).
    if (texcoord >= uid_data->genMode_numtexgens)
      texcoord = 0;

    out->Write("\tcustom_data.texmap_to_texcoord_index[{}] = {};\n", texmap, texcoord);
  }
}
```

You'd probably want to do this before the current code that uses `tevstage.tevorders_texcoord` so that those values overwrite the previous values (though I don't think there is ever a situation where using a texture both as an indirect texture (what the videos referred to as a displacement map in some cases) and also as a source for colors to put on screen would make sense).

Reply:
Thank you!!  I came to the same conclusion.  I pushed that up, I did leave the tex coord modification after looking at Dolphin's logic a bit more.  Seems I can avoid the scale problem by having a separate variable.

Unfortunately, graphics mods don't (easily) work with fifo logs.  So I might punt on that until I get that investigation until I finish my editor.

Reply:
Added a comment with your updated findings.  Thank you!